docs/design/fused_moe_modular_kernel.md

```diff
@@ -140,19 +150,32 @@ FusedMoEPermuteExpertsUnpermute performs the core of the FusedMoE operations. Th
 `FusedMoEPermuteExpertsUnpermute::apply`: Refer to `FusedMoEPermuteExpertsUnpermute` section above.
 
 ### FusedMoEModularKernel Initialization
-During Engine startup,
-* The `FusedMoEPrepareAndFinalize` object is created in [vllm/model_executor/layers/fused_moe/layer.py](https://github.com/vllm-project/vllm/blob/5ac3168ee342f4cae17b0b67375e647bd5dd9151/vllm/model_executor/layers/fused_moe/layer.py#L190)
-* The `FusedMoEPermuteExpertsUnpermute` object is created in `FusedMoEMethodBase::select_gemm_impl`
+`FusedMoEMethodBase` class has 2 methods that are collectively responsible in creating the `FusedMoEModularKernel` object. They are,
+* select_gemm_impl, and
+* init_prepare_finalize
+
+#### select_gemm_impl
+The `select_gemm_impl` method is undefined in the base class. It is the responsibility of the derived class to implement a method that constructs a valid/appropriate `FusedMoEPermuteExpertsUnpermute` object.
+Please refer to the implementations in,
+* `UnquantizedFusedMoEMethod`
+* `CompressedTensorsW8A8Fp8MoEMethod`
+* `CompressedTensorsW8A8Fp8MoECutlassMethod`
+* `Fp8MoEMethod`
+* `ModelOptNvFp4FusedMoE`
+dervied classes.
+
+#### init_prepare_finalize
+Based on the input and env settings, the `init_prepare_finalize` method creates the appropriate `FusedMoEPrepareAndFinalize` object. The method then queries `select_gemm_impl` for the appropriate `FusedMoEPermuteExpertsUnpermute` object and builds the `FusedMoEModularKernel` object
```

Comment:
I think it's important to note that layer.py sets `self.fused_experts` to the fused moe object and that the subclass needs to use this in it's own `apply` method.